File: javascript/test/bip39.test.js
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { test } from 'node:test';
import {
  entropyToMnemonic,
  generateMnemonic,
  mnemonicToSeedHex,
  parseMnemonic,
  validateMnemonic
} from '../index.js';

const vectors = JSON.parse(readFileSync(new URL('./english-vectors.json', import.meta.url), 'utf8'));

test('matches official English BIP-39 vectors', () => {
  for (const vector of vectors) {
    assert.equal(entropyToMnemonic(vector.entropyHex), vector.mnemonic);
    assert.equal(parseMnemonic(vector.mnemonic), vector.mnemonic);
    assert.equal(mnemonicToSeedHex(vector.mnemonic, 'TREZOR'), vector.seedHex);
  }
});

test('generates valid mnemonics', () => {
  for (const words of [12, 15, 18, 21, 24]) {
    const mnemonic = generateMnemonic({ words });
    assert.equal(mnemonic.split(' ').length, words);
    assert.equal(validateMnemonic(mnemonic), true);
  }
});

test('rejects invalid checksum', () => {
  assert.equal(validateMnemonic('abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon'), false);
});

File: dart/test/bip39_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:bip39_mnemonic/bip39_mnemonic.dart';
import 'package:test/test.dart';

void main() {
  test('matches official English vectors', () {
    final vectors = jsonDecode(File('dart/test/english-vectors.json').readAsStringSync()) as List<dynamic>;
    for (final vector in vectors.cast<Map<String, dynamic>>()) {
      expect(entropyHexToMnemonic(vector['entropyHex'] as String), vector['mnemonic']);
      expect(parseMnemonic(vector['mnemonic'] as String), vector['mnemonic']);
      expect(mnemonicToSeedHex(vector['mnemonic'] as String, passphrase: 'TREZOR'), vector['seedHex']);
    }
  });

  test('generates valid mnemonics', () {
    for (final count in [12, 15, 18, 21, 24]) {
      final mnemonic = generateMnemonic(words: count);
      expect(mnemonic.split(' '), hasLength(count));
      expect(validateMnemonic(mnemonic), isTrue);
    }
  });
}
